internal/databases/greenplum/follow_primary_handler.go

```diff
@@ -0,0 +1,171 @@
+		cluster:            globalCluster,
+		stopAtRestorePoint: stopAtRestorePoint,
+		timeoutInSeconds:   timeoutInSeconds,
+	}
+}
+
+func FatalIfWalLogMissing(restorePoint string, folder storage.Folder) {
+	metadata, err := FetchRestorePointMetadata(folder, restorePoint)
+	if err != nil {
+		tracelog.ErrorLogger.FatalOnError(err)
+	}
+
+	for seg, lsn := range metadata.LsnBySegment {
+		LSN, err := postgres.ParseLSN(lsn)
+		if err != nil {
+			tracelog.ErrorLogger.FatalOnError(err)
+		}
+		walSegmentNo := postgres.NewWalSegmentNo(LSN)
+
+		subfolder := folder.GetSubFolder(fmt.Sprintf(WalFolder, seg))
+		folderObjects, _, err := subfolder.ListFolder()
+		if err != nil {
+			tracelog.ErrorLogger.FatalOnError(err)
+		}
+
+		// WAL file example: "000000010000000000000003.lz4" -> base name is "000000010000000000000003"
+		found := false
+		for _, obj := range folderObjects {
+			if strings.HasPrefix(obj.GetName(), walSegmentNo.GetFilename(metadata.TimeLine)) {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			tracelog.ErrorLogger.Fatalln("WAL file was not uploaded for all segments and master")
+		}
+	}
+}
+
+func (fh *FollowPrimaryHandler) Follow() {
+	tracelog.InfoLogger.Println("Updating recovery.conf on segments and master...")
+	fh.updateRecoveryConfigs()
+	fh.applyXLogInCluster()
+}
+
+func (fh *FollowPrimaryHandler) applyXLogInCluster() {
+	tracelog.InfoLogger.Println("Running recovery on segments and master...")
+	// Run WAL-G to restore the each segment as a single Postgres instance
+	remoteOutput := fh.cluster.GenerateAndExecuteCommand("Running wal-g",
+		cluster.ON_SEGMENTS|cluster.INCLUDE_MASTER,
+		func(contentID int) string {
+			return fh.buildSegmentStartCommand(contentID)
+		})
+
+	fh.cluster.CheckClusterError(remoteOutput, "Unable to run wal-g", func(contentID int) string {
+		return "Unable to run wal-g"
+	})
+
+	for _, command := range remoteOutput.Commands {
+		tracelog.DebugLogger.Printf("WAL-G output (segment %d):\n%s\n", command.Content, command.Stderr)
+	}
+}
+
+// updateRecoveryConfigs generates and uploads the correct recovery.conf
+// files to each segment instance (including master) so they can recover correctly
+// up to the expected restore point
+func (fh *FollowPrimaryHandler) updateRecoveryConfigs() {
+	recoveryTarget := fh.stopAtRestorePoint
+	tracelog.InfoLogger.Printf("Recovery target is %s", recoveryTarget)
+	restoreCfgMaker := NewRecoveryConfigMaker(
+		"wal-g", conf.CfgFile, recoveryTarget, true)
+
+	remoteOutput := fh.cluster.GenerateAndExecuteCommand("Updating recovery.conf on segments and master",
+		cluster.ON_SEGMENTS|cluster.INCLUDE_MASTER,
+		func(contentID int) string {
+			segment := fh.cluster.ByContent[contentID][0]
+			pathToRestore := path.Join(segment.DataDir, viper.GetString(conf.GPRelativeRecoveryConfPath))
+			// For this feature, we expect Cloudberry / Greenplum 6.25+ (in this version some patches from 9.5 were backported)
+			fileContents := restoreCfgMaker.Make(contentID, 90500)
```

Comment:
It is better to have some way to detect features supported by gpdb/cbdb... =/ 
Will think a bit.

Reply:
Probably, we should do nothing with this. Just wait untill greenplum 6.24- phase out.